### PR TITLE
Enable VIT ReciproCAM

### PR DESCRIPTION
**Pull request by negvet**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/hooks/recording_forward_hook.py [hunk 1]
```diff
@@ -179,6 +179,116 @@ def _get_mosaic_feature_map(self, feature_map: torch.Tensor, c: int, h: int, w:
...
+    """Implementation of ViTRecipro-CAM for class-wise saliency map for transformer-based classifiers.
+
+    Args:
+    head_forward_fn (callable): Forward pass function for the top of the model.
+    num_classes (int): Number of classes.
+    use_gaussian (bool): Defines kernel type for mosaic feature map generation.
+    If True, use gaussian 3x3 kernel. If False, use 1x1 kernel.
+    cls_token (bool): If True, includes classification token into the mosaic feature map.
+    normalize (bool): If True, Normalizes saliency maps.
+    """
```
Comment (GalyaZalesskaya) [inline src/otx/algo/hooks/recording_forward_hook.py]:
```suggestion
    Args:
        head_forward_fn (callable): Forward pass function for the top of the model.
        num_classes (int): Number of classes.
        use_gaussian (bool): Defines kernel type for mosaic feature map generation.
            If True, use gaussian 3x3 kernel. If False, use 1x1 kernel.
        cls_token (bool): If True, includes classification token into the mosaic feature map.
        normalize (bool): If True, Normalizes saliency maps.
    """
```

File: src/otx/algo/hooks/recording_forward_hook.py [hunk 2]
```diff
@@ -179,6 +179,116 @@ def _get_mosaic_feature_map(self, feature_map: torch.Tensor, c: int, h: int, w:
...
+                mosaic_feature_map_mask[:, 0] = torch.ones(1).to(feature_map.device)
+            mosaic_feature_map_mask = torch.tensor(mosaic_feature_map_mask.unsqueeze(2).repeat(1, 1, dim))
+            mosaic_feature_map = feature_map_repeated * mosaic_feature_map_mask
+
+        return mosaic_feature_map
+
+    def _predict_from_feature_map(self, x: torch.Tensor) -> torch.Tensor:
+        with torch.no_grad():
+            logits = self._head_forward_fn(x)
+            if not isinstance(logits, torch.Tensor):
```
Comment (GalyaZalesskaya) [inline src/otx/algo/hooks/recording_forward_hook.py]: Let's move this function to `BaseRecordingForwardHook` and remove it from here and `ReciproCAMHook`. And I'll update `DetClassProbabilityMapHook` respectively

**Review by GalyaZalesskaya — COMMENTED** [review]
Can you please also extend the scope of e2e tests after this update, fixing  [test_cli.py:L154](https://github.com/openvinotoolkit/training_extensions/blob/d86c4f0680522b931558c92a112488dd3f994952/tests/integration/cli/test_cli.py#L154)?

File: src/otx/algo/classification/deit_tiny.py [hunk 3]
```diff
@@ -2,17 +2,70 @@
...
+        self.explain_hook = ViTReciproCAMHook.create_and_register_hook(
+            target_layernorm,
+            self.head_forward_fn,
+            num_classes=self.num_classes,
+        )
+
+    def get_target_layernorm(self, final_norm: bool = True) -> torch.nn.Module:
+        """Returns the first (out of two) layernorm layer from the last backbone layer."""
+        layernorm_layers = [module for module in self.backbone.modules() if isinstance(module, torch.nn.LayerNorm)]
+        target_layernorm_index = -2 - int(final_norm)
```
Comment (sovrasov) [inline src/otx/algo/classification/deit_tiny.py:35]: Is that for Deit only?

File: src/otx/algo/hooks/recording_forward_hook.py (hunk 1 above)
Comment (negvet) [reply src/otx/algo/hooks/recording_forward_hook.py]: Fixed

File: src/otx/algo/hooks/recording_forward_hook.py (hunk 2 above)
Comment (negvet) [reply src/otx/algo/hooks/recording_forward_hook.py]: Good point, thanks. Done

**Review by sovrasov — APPROVED** [review]

**Review by GalyaZalesskaya — APPROVED** [review]

**Review by vinnamkim — APPROVED** [review]

**Review by MarkByun — APPROVED** [review]
LGTM

**Pull request merged**